pkg/ccl/backupccl/backup_metadata.go

```diff
@@ -0,0 +1,1135 @@
+	return DescriptorRevisionIterator{
+		backing: backing,
+	}
+}
+
+// Close closes the iterator.
+func (dri *DescriptorRevisionIterator) Close() {
+	dri.backing.close()
+}
+
+// Err returns the iterator's error.
+func (dri *DescriptorRevisionIterator) Err() error {
+	if dri.err != nil {
+		return dri.err
+	}
+	return dri.backing.err()
+}
+
+// Next retrieves the next descriptor revision in the iterator.
+//
+// Next returns true if next element was successfully unmarshalled into
+// revision, and false if there are no more elements or if an error was
+// encountered. When Next returns false, the user should call the Err method to
+// verify the existence of an error.
+func (dri *DescriptorRevisionIterator) Next(revision *BackupManifest_DescriptorRevision) bool {
+	wrapper := resultWrapper{}
+	ok := dri.backing.next(&wrapper)
+	if !ok {
+		return false
+	}
+
+	err := unmarshalWrapper(&wrapper, revision)
+	if err != nil {
+		dri.err = err
+		return false
+	}
+
+	return true
+}
+
+func unmarshalWrapper(wrapper *resultWrapper, rev *BackupManifest_DescriptorRevision) error {
+	var desc *descpb.Descriptor
+	if len(wrapper.value) > 0 {
+		desc = &descpb.Descriptor{}
+		err := protoutil.Unmarshal(wrapper.value, desc)
+		if err != nil {
+			return err
+		}
+	}
+
+	id, err := decodeDescSSTKey(wrapper.key.Key)
+	if err != nil {
+		return err
+	}
+
+	*rev = BackupManifest_DescriptorRevision{
+		Desc: desc,
+		ID:   id,
+		Time: wrapper.key.Timestamp,
+	}
+	return nil
+}
+
+// StatsIterator is a simple iterator to iterate over stats.TableStatisticProtos.
+type StatsIterator struct {
+	backing bytesIter
+	err     error
+}
+
+// StatsIter creates a new StatsIterator for the backup metadata.
+func (b *BackupMetadata) StatsIter(ctx context.Context) StatsIterator {
+	backing := makeBytesIter(ctx, b.store, b.filename, []byte(sstStatsPrefix), b.enc, false)
+	return StatsIterator{
+		backing: backing,
+	}
+}
+
+// Close closes the iterator.
+func (si *StatsIterator) Close() {
+	si.backing.close()
+}
+
+// Err returns the iterator's error.
+func (si *StatsIterator) Err() error {
+	if si.err != nil {
+		return si.err
+	}
+	return si.backing.err()
+}
+
+// Next retrieves the next stats proto in the iterator.
+//
+// Next returns true if next element was successfully unmarshalled into
+// statsPtr, and false if there are no more elements or if an error was
+// encountered. When Next returns false, the user should call the Err method to verify the
+// existence of an error.
+func (si *StatsIterator) Next(statsPtr **stats.TableStatisticProto) bool {
+	wrapper := resultWrapper{}
+	ok := si.backing.next(&wrapper)
+
+	if !ok {
+		return false
+	}
+
+	var s stats.TableStatisticProto
+	err := protoutil.Unmarshal(wrapper.value, &s)
+	if err != nil {
+		si.err = err
+		return false
+	}
+
+	*statsPtr = &s
+	return true
+}
+
+type bytesIter struct {
+	Iter storage.SimpleMVCCIterator
+
+	prefix      []byte
+	useMVCCNext bool
+	iterError   error
+}
+
+func makeBytesIter(
```

Comment:
I'm probably missing something, but it seems like we already open the SST in makeMetadata, to read the fixed-set of details (nee manifest) so we could just pass that one instead of re-opening it for each of these typed iterators? 

And then perhaps the typed iterators, if they just seek and read from it directly, could potentially also avoid some cloning of unsafe value just to pass it to unmarshal anyway, so maybe we wouldn't even need the intermedia bytesIter at all?

Reply:
I was trying to avoid the case where someone might use multiple of these iterators in a nested manner, or instantiate another iterator without exhausting the the current one, e.g. in something like this where spans is nested inside descriptors: https://github.com/cockroachdb/cockroach/blob/4f4f2accb4a26f89032a4587b07e788cc440c08c/pkg/ccl/backupccl/restore_planning.go#L1915

Reply:
We can Seek() the iterator around to different key ranges if we want to read some things here, then jump to some there, and if someone knows they really want to consume a little this and a little that side-by-side at the same time, they could choose themselves to just open duplicate iterators to avoid the seeks, so I don't know if I'd prescribe that ahead of time by forcing creation of type-specific readers.